File: python/src/p_0027_remove_element.py
# https://leetcode.com/problems/remove-element/

from typing import List

class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        k = 0
        for i in range(len(nums)):
            if nums[i] != val:
                nums[k] = nums[i]
                k += 1
        return k



def tests():
    s = Solution()

    nums = [3, 2, 2, 3]
    val = 3
    k = s.removeElement(nums, val)
    print(f'nums={nums}, val={val}, k={k}')
    assert k == 2
    assert nums[:k] == [2, 2]
    print('-' * 10)

    nums = [0,1,2,2,3,0,4,2]
    val = 2
    k = s.removeElement(nums, val)
    print(f'nums={nums}, val={val}, k={k}')
    assert k == 5
    assert nums[:k] == [0,1,3,0,4]
    print('-' * 10)

    nums = []
    val = 0
    k = s.removeElement(nums, val)
    print(f'nums={nums}, val={val}, k={k}')
    assert k == 0
    assert nums[:k] == []
    print('-' * 10)

    nums = [1]
    val = 1
    k = s.removeElement(nums, val)
    print(f'nums={nums}, val={val}, k={k}')
    assert k == 0
    assert nums[:k] == []
    print('-' * 10)

    nums = [1]
    val = 0
    k = s.removeElement(nums, val)
    print(f'nums={nums}, val={val}, k={k}')
    assert k == 1
    assert nums[:k] == [1]
    print('-' * 10)


if __name__ == '__main__':
    tests()

File: typescript/src/p_0027_remove_element.ts
// https://leetcode.com/problems/remove-element/

import assert from "node:assert";

function removeElement(nums: number[], val: number): number {
    let k = 0;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] !== val) {
            nums[k] = nums[i];
            k++;
        }
    }
    return k;
};


function tests() {
    let nums = [3, 2, 2, 3];
    let val = 3;
    let expected = 2;
    let result = removeElement(nums, val);
    assert.strictEqual(result, expected, `Test case 1 failed: removeElement([${nums}], ${val}) = ${result}, expected ${expected}`);
    console.log(`Test case 1 passed: removeElement([${nums}], ${val}) = ${result}`);

    nums = [0,1,2,2,3,0,4,2];
    val = 2;
    expected = 5;
    result = removeElement(nums, val);
    assert.strictEqual(result, expected, `Test case 2 failed: removeElement([${nums}], ${val}) = ${result}, expected ${expected}`);
    console.log(`Test case 2 passed: removeElement([${nums}], ${val}) = ${result}`);

    nums = [1];
    val = 1;
    expected = 0;
    result = removeElement(nums, val);
    assert.strictEqual(result, expected, `Test case 3 failed: removeElement([${nums}], ${val}) = ${result}, expected ${expected}`);
    console.log(`Test case 3 passed: removeElement([${nums}], ${val}) = ${result}`);
}

tests();

